Add `relativedelta` testing

2br-2b opened the pull request:
I'm not sure if this is the proper way to contribute to your PR #6, but I added in unit tests to ensure that the conversion is working properly.

2br-2b committed:
Split tests into different classes

2br-2b committed:
Added relativedelta tests

2br-2b commented:
The only test I'm uncertain about is:

```
======================================================================
FAIL: test_combined (tests.TestRelativeDeltaOutput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\abold\Desktop\pytimeparse2\tests.py", line 769, in test_combined
    self.assertEqual(timeparse.parse('1y2mo3w4d5h6m7s8ms', as_timedelta=True), relativedelta(seconds = 38898367, microseconds=8000))
AssertionError: relativedelta(years=+1, months=+2, days=+25, hours=+5, [38 chars]8000) != relativedelta(days=+450, hours=+5, minutes=+6, seconds=[18 chars]8000)
```

In theory, conversion should work fine (`relativedelta(minutes = 1) == relativedelta(seconds = 60)`), but given that the second option lists days as 450 rather than splitting it into years etc., something could be going wrong there.

2br-2b committed:
Removed extra lines to make regex matching easier

2br-2b committed:
Simplified single assertEqual case

2br-2b commented:
Looking at these tests:
```python
    def test_timeparse_unparsed(self):
        """Check that unparsed values tries to converts into int(). """
        self.assertEqual(timeparse.parse(100), 100)
        self.assertEqual(timeparse.parse(-18.333), -18)
        self.assertEqual(timeparse.parse('99.1'), 99)
        self.assertEqual(timeparse.parse('-99.1'), -99)
```
and 
```python
    def test_plain_numbers(self):
        self.assertEqual(timeparse.parse('10'), 10)
        self.assertEqual(timeparse.parse('10.1'), 10)
        self.assertEqual(timeparse.parse('-10'), -10)
        self.assertEqual(timeparse.parse('-10.1'), -10)
```

Do we want to leave these as is? If we're returning floats, then it seems these would need to be corrected. What is the purpose of these tests, and do we want to maintain that?

2br-2b committed:
Corrected relativedelta test parameters
See https://github.com/dateutil/dateutil/issues/1282#issue-1700291981

onegreyonewhite commented:
> Looking at these tests:
> 
> ```python
>     def test_timeparse_unparsed(self):
>         """Check that unparsed values tries to converts into int(). """
>         self.assertEqual(timeparse.parse(100), 100)
>         self.assertEqual(timeparse.parse(-18.333), -18)
>         self.assertEqual(timeparse.parse('99.1'), 99)
>         self.assertEqual(timeparse.parse('-99.1'), -99)
> ```
> 
> and
> 
> ```python
>     def test_plain_numbers(self):
>         self.assertEqual(timeparse.parse('10'), 10)
>         self.assertEqual(timeparse.parse('10.1'), 10)
>         self.assertEqual(timeparse.parse('-10'), -10)
>         self.assertEqual(timeparse.parse('-10.1'), -10)
> ```
> 
> Do we want to leave these as is? If we're returning floats, then it seems these would need to be corrected. What is the purpose of these tests, and do we want to maintain that?

Yes, we need this. First, it is necessary to maintain backward compatibility. Secondly, my policy is that if milliseconds are not explicitly specified, then an integer should be returned. This is in the style of Python, because if you do not specify a number through a dot, then you get an integer.

onegreyonewhite commented on tests.py:
```diff
@@ -108,6 +109,17 @@ def test_time(self):
+
+    def setUp(self):
+        """Setup function."""
+        pass
```
@2br-2b why is it needed?

2br-2b replied:
I assumed they were needed, as they were in the original changes. Since it appears they aren't, I'll remove them.

onegreyonewhite commented on tests.py:
```diff
@@ -269,14 +281,11 @@ def test_timeparse_15(self):
             192896)
-        self.assertEqual(
-            timeparse.parse('-2 days, 5 hours, 34 minutes, 56 seconds'),
+        self.assertEqual(timeparse.parse('-2 days, 5 hours, 34 minutes, 56 seconds'),
```
@2br-2b why this?

2br-2b replied:
It must have been changed when I was removing the other line breaks to make regexing easier. I could undo these changes if you like

onegreyonewhite replied:
@2br-2b I hate to be "that guy", but can u return reverse changes?) It more readable for some cases.

2br-2b replied:
No problem at all! I just fixed it in 6de5b8b054f5a0155215785b7f6f078ae7e6eeea

onegreyonewhite commented on tests.py:
```diff
@@ -426,6 +435,327 @@ def test_plain_numbers(self):
+        self.assertEqual(timeparse.parse('-5 day', as_timedelta=True), relativedelta(seconds = -432000))
+
+    def test_timeparse_30(self):
+        """timeparse test case 30."""
+        self.assertEqual(timeparse.parse('5 days', as_timedelta=True), relativedelta(seconds = 432000))
+        self.assertEqual(timeparse.parse('+5 days', as_timedelta=True), relativedelta(seconds = 432000))
+        self.assertEqual(timeparse.parse('-5 days', as_timedelta=True), relativedelta(seconds = -432000))
+
+    def test_timeparse_31(self):
+        """timeparse test case 31."""
+        self.assertEqual(timeparse.parse('5.6 wk', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('+5.6 wk', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('-5.6 wk', as_timedelta=True), relativedelta(seconds = -3386880))
+
+    def test_timeparse_32(self):
+        """timeparse test case 32."""
+        self.assertEqual(timeparse.parse('5.6 week', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('+5.6 week', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('-5.6 week', as_timedelta=True), relativedelta(seconds = -3386880))
+
+    def test_timeparse_33(self):
+        """timeparse test case 33."""
+        self.assertEqual(timeparse.parse('5.6 weeks', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('+5.6 weeks', as_timedelta=True), relativedelta(seconds = 3386880))
+        self.assertEqual(timeparse.parse('-5.6 weeks', as_timedelta=True), relativedelta(seconds = -3386880))
+
+    def test_milliseconds(self):
+        self.assertEqual(timeparse.parse('3 ms', as_timedelta=True), relativedelta(microseconds= 3000))
+        self.assertEqual(timeparse.parse('3 millis', as_timedelta=True), relativedelta(microseconds = 3000))
+        self.assertEqual(timeparse.parse('3 msecs', as_timedelta=True), relativedelta(microseconds = 3000))
+        self.assertEqual(timeparse.parse('3 milliseconds', as_timedelta=True), relativedelta(microseconds = 3000))
+
+    def test_plain_numbers(self):
+        self.assertEqual(timeparse.parse('10', as_timedelta=True), relativedelta(seconds = 10))
+        self.assertEqual(timeparse.parse('10.1', as_timedelta=True), relativedelta(seconds = 10, microseconds = 100000))
+        self.assertEqual(timeparse.parse('-10', as_timedelta=True), relativedelta(seconds = -10))
+        self.assertEqual(timeparse.parse('-10.1', as_timedelta=True), -relativedelta(seconds = 10, microseconds = 100000))
+
+    def test_combined(self):
+        self.assertEqual(timeparse.parse('1y2mo3w4d5h6m7s8ms', as_timedelta=True), relativedelta(seconds = 38898367, microseconds=8000))
```
@2br-2b good work! I think it would be better to refactor this. `TestRelativeDeltaOutput` need just one test method which call some patterned method for each value and check equality of two values:
- value of constant date minus value of `parse(..., as_timedelta=True)`
- value of constant date minus value of `relativedelta(seconds=parse(...))`

What do you think?

onegreyonewhite replied:
@2br-2b I've just added support for disabling ``dateutil`` support, because that trick wouldn't work with it.

2br-2b replied:
> value of constant date minus value of `relativedelta(seconds=parse(...))`
The only problem is that this won't work with all of the values. For example, `relativedelta(days=365) != relativedelta(years=1)` because some years are leap years. Thus, using relativedelta for parsing months and years won't work. On that note, I think I need to update some of these tests...

I do agree that we could create one function to run the tests; it could take a string and a relativedelta as input and output a boolean.

I'm realizing now; should we have three suites of tests:
1. integer/float output
2. relativedelta output
3. datetime output

? If so, it might be good to abstract the `TestRelativeDeltaOutput` class; on the other hand, I don't want to fall into a trap of over-abstracting everything

2br-2b committed:
Remove setUp functions
See github.com/onegreyonewhite/pytimeparse2/pull/7#discussion_r1189229091

2br-2b committed:
Refactored relativedelta tests

2br-2b committed:
Refactor `test_timeparse_9`
Removed spaces around multiply symbol to make it consistent with other
test cases

2br-2b commented:
> Secondly, my policy is that if milliseconds are not explicitly specified, then an integer should be returned. This is in the style of Python, because if you do not specify a number through a dot, then you get an integer.

Since I don't think I quite understand this point, it might be best for you to go thru the changes in `TestRelativeDeltaOutput` and remove `microseconds` from the cases where they don't belong.

onegreyonewhite commented:
> > Secondly, my policy is that if milliseconds are not explicitly specified, then an integer should be returned. This is in the style of Python, because if you do not specify a number through a dot, then you get an integer.
> 
> Since I don't think I quite understand this point, it might be best for you to go thru the changes in `TestRelativeDeltaOutput` and remove `microseconds` from the cases where they don't belong.

Ok.

> If so, it might be good to abstract the TestRelativeDeltaOutput class; on the other hand, I don't want to fall into a trap of over-abstracting everything.

Ok. Just let me know when you're done.

onegreyonewhite commented:
@2br-2b please, dont forget to pull from branch.

2br-2b commented:
<details>
<summary>Simple git question 😅</summary>

> @2br-2b please, dont forget to pull from branch.

Really sorry to bother you with such a simple request, but do I merge or rebase my branch? I'm not too familiar with this specifically, and every time I've tried in the past, I end up with duplicate commits. I believe after I pull, I run:
```
git checkout relativedelta-testing
git merge feature/timedelta-support
git push
```
correct? I tried that, but I think I was causing some duplicate commits.

The other option would be:
```
git checkout relativedelta-testing
git rebase feature/timedelta-support
git push
```
I just want to be sure I don't mess up the commit log of this project. Thanks for the help 😅
</details>

2br-2b committed:
Fix formatting of `test_timeparse_16`

onegreyonewhite commented:
@2br-2b you need just call `git pull https://github.com/onegreyonewhite/pytimeparse2.git feature/timedelta-support` from your branch without any rebase or commits. It's the easiest way now)) This action pull all missed commits. You don't have any conflicts in changes. New merge commit ok, because I'm going to rebase all PR before merge it (standard gh functionality).

P.S.: dont forget push it))

2br-2b committed:
Merge branch 'feature/timedelta-support' of https://github.com/onegreyonewhite/pytimeparse2 into relativedelta-testing

onegreyonewhite reviewed (approved):
Awesome🙌🏼
